Sources/NIOHTTPClient/HTTPHandler.swift

```diff
@@ -19,24 +19,40 @@ import NIOHTTP1
 import NIOSSL
 
 public extension HTTPClient {
-    enum Body: Equatable {
-        case byteBuffer(ByteBuffer)
-        case data(Data)
-        case string(String)
-
-        var length: Int {
-            switch self {
-            case .byteBuffer(let buffer):
-                return buffer.readableBytes
-            case .data(let data):
-                return data.count
-            case .string(let string):
-                return string.utf8.count
+    struct Body {
+        typealias ChunkProvider = (@escaping (IOData) -> EventLoopFuture<Void>) -> EventLoopFuture<Void>
+
+        var length: Int?
+        var provider: ChunkProvider
+
+        static func byteBuffer(_ buffer: ByteBuffer) -> Body {
+            return Body(length: buffer.readableBytes) { writer in
+                writer(.byteBuffer(buffer))
+            }
+        }
+
+        static func stream(length: Int? = nil, _ provider: @escaping ChunkProvider) -> Body {
+            return Body(length: length, provider: provider)
+        }
+
+        static func data(_ data: Data) -> Body {
+            return Body(length: data.count) { writer in
+                var buffer = ByteBufferAllocator().buffer(capacity: data.count)
+                buffer.writeBytes(data)
```

Comment:
`HTTPPart.body` requires `IOData`, and `IOData` could be `.byteBuffer` or `.fileRegion`, so we have to write to a byte buffer. I'm not sure there is a way to avoid copying here, @weissi is this correct?

Reply:
@artemredkin is right, there's no way to do this.

Reply:
Shame. Could we teach `IOData` about `Data`?

Reply:
We could, but it won't help much.

Our underlying writing mechanism heavily uses the reader/writer indices of `ByteBuffer`. This means we need a `ByteBuffer` *eventually*, which means we need to copy *eventually*. At least in this case we can do it as close to the user as possible, making it clearer in profiling, rather than burying that copy deep in NIO's I/O loop.